Add aoc-linux-fixes tool, modified multiprocessing for efficiency

dakotaSicher opened the pull request:
- Fixed commit.hex to str(commit.id)
- Created aoc-extract-linux fixes tool
   -moved main function of extract_fixes.py to aoc-linux-fixes.py in tools
   -added aoc-linux-fixes to setup.py for install
   -added command line options for linux directory, output directory, multiprocessing, and history length
   -added ability to select length of time for commits to process instead of always going all the way back to the when "Fixes:" was introduced.
- modified multiprocessing for linux fixes to use a shared queue instead of static partitioning of commits

- added new tool to readme 
- added additional dependencies to setup.py

dakotaSicher committed:
added cli for linux fixes

dakotaSicher committed:
moved aoc_linux_fixes to tools

dakotaSicher committed:
Merge pull request #1 from dakotaSicher/wip

added new command line tool for running linux_fixes, updated required dependencies, fixed commit.hex issue

dakotaSicher committed:
added number of process to cmd line tool for linux fixes

dakotaSicher committed:
changed multiprocessing mode for processing linux commits from using static division of commits for each process to a shared queue

dakotaSicher committed:
removed extra argument from _get_removed_lines_worker. added profiling output files to gitignore

dakotaSicher committed:
Merge remote-tracking branch 'origin/wip'

dakotaSicher committed:
Merge pull request #2 from dakotaSicher/wip

Wip

dakotaSicher committed:
Updated README with aoc-linux-fixes tool. removed redundant extract_linux_fixes function in extract_fixes.py. Added comments to aoc-linux-fixes.py for usage. other comments

dakotaSicher committed:
removed '-2' from the cpu count used for testing

renatav commented on src/analysis/scripts/extract_fixes.py:
```diff
@@ -15,19 +18,19 @@
-LAST_PROCESSED_DIR = Path("../last_processed")
-ERRORS_FILE_DIR = RESULTS_DIR/ "extract"
-NUMBER_OF_PROCESSES = 5
+REPO_PATH = (ROOT_DIR.parent / "projects/linux").absolute()  # Path to the Linux kernel Git repository
```
Ideally, this should be configurable. We could either have a conf file that would be added to gitignore where a developer would set this path, or specified using the CLI- or both, where if specified using the CLI, that value would take precedence over paths defined elsewhere

dakotaSicher replied:
i just forgot to remove these, they are no longer used in this file. removed

renatav commented on src/analysis/scripts/extract_fixes.py:
```diff
@@ -130,14 +149,18 @@ def get_removed_lines(repo_path, commits, output, processed_path, errors_path):
     """
     Main function to spawn the processes.
     """
+    if number_of_processes == 0:
```
Can be just `not number_of_processes`, so will be True both if  `number_of_processes` is 0 and is `None`

dakotaSicher replied:
good idea, change made

renatav commented on src/run_cocci.py:
```diff
@@ -152,7 +152,16 @@ def postprocess_and_generate_output(file_path: Path,  patch: CocciPatch, remove_
     return filtered_data
 
 
-def run_patches_and_generate_output(input_path: Path, output_path: Optional[Path] = None, temp_dir: Optional[Path] = None, split_output = True, patch: Optional[CocciPatch] = None,  patches_to_skip: Optional[list] = None, remove_end_line_and_col: Optional[bool]=True, cocci_dir:Optional[Path]=COCCI_DIR):
```
Thanks for formatting this. Do you have any tool set up on your end? We used Black in the past and it's pretty good. We could even configure it to run on CI on push - if getting some experience with CI and formatting tools is something you are interested in

dakotaSicher replied:
I just manually edited so i could read the whole thing. I will take a look at Black and CI stuff, never done that before but sound like a very useful thing to know.

renatav commented on src/tools/aoc_linux_fixes.py:
```diff
@@ -0,0 +1,66 @@
+@click.argument("linux_dir",type=Path)
+@click.option("-o", "--output-dir", type=Path, default="./output")
+@click.option("-t","--history_length",type=str, default=None,help="Length of commit history to analyze")
+@click.option("-p","--cpus",type=int,default=1,help="number of cpus to use for multiprocessing. Enter 0 to select all")
```
Could be optional and set to all if not specified

dakotaSicher replied:
that sounds more intuitive

renatav reviewed (commented):
Nice work! I haven't tested on my machine, but the code looks good. Just some minor comments

dakotaSicher committed:
Update extract_fixes.py

removed unused constants, updated number_of_processes check to include none in execute and execute_queue

dakotaSicher committed:
Update aoc_linux_fixes.py

updated cpus option to default to none, which will later result in cpu_count being used in execute_queue
